File: app/src/main/java/com/pixplicity/cryptogram/views/HintView.java
package com.pixplicity.cryptogram.views;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatTextView;
import android.text.TextPaint;
import android.util.AttributeSet;

import com.pixplicity.cryptogram.R;
import com.pixplicity.cryptogram.models.Puzzle;
import com.pixplicity.cryptogram.utils.StyleUtils;

import java.util.Collection;


public class HintView extends AppCompatTextView {

    private static final String TAG = HintView.class.getSimpleName();

    @Nullable
    private Puzzle mPuzzle;

    private int mCharsPerRow = 1;

    private float mMinBoxW, mBoxW, mCharH, mCharW;
    private TextPaint mTextPaint;


    public HintView(Context context) {
        super(context);
        init(context, null, 0);
    }

    public HintView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs, 0);
    }

    public HintView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        Resources res = context.getResources();

        mTextPaint = new TextPaint();
        mTextPaint.setColor(getCurrentTextColor());
        mTextPaint.setAntiAlias(true);
        mTextPaint.setTypeface(Typeface.MONOSPACE);

        // Compute size of each box
        mMinBoxW = StyleUtils.getSize(res, R.dimen.puzzle_box_width);
        mTextPaint.setTextSize(StyleUtils.getSize(res, R.dimen.puzzle_hint_size));

        // Compute size of a single char (assumes monospaced font!)
        Rect bounds = new Rect();
        mTextPaint.getTextBounds("M", 0, 1, bounds);
        mCharW = bounds.width();
        mCharH = bounds.height();

        if (isInEditMode()) {
            setPuzzle(new Puzzle());
        }
    }

    @Nullable
    public Puzzle getPuzzle() {
        return mPuzzle;
    }

    public void setPuzzle(Puzzle puzzle) {
        mPuzzle = puzzle;
        requestLayout();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);

        int desiredWidth = getSuggestedMinimumWidth();

        int width;
        //Measure Width
        if (widthMode == MeasureSpec.EXACTLY) {
            //Must be this size
            width = widthSize;
        } else if (widthMode == MeasureSpec.AT_MOST) {
            //Can't be bigger than...
            width = Math.min(desiredWidth, widthSize);
        } else {
            //Be whatever you want
            width = desiredWidth;
        }

        // Pack the most number of characters in the bar
        int innerBox = width - getPaddingLeft();
        for (int i = 1; i < 26; i++) {
            mCharsPerRow = (int) Math.ceil(26f / i);
            mBoxW = innerBox / mCharsPerRow;
            if (mBoxW >= mMinBoxW) {
                break;
            }
        }

        int desiredHeight = drawChars(null, width);

        int height;
        //Measure Height
        if (heightMode == MeasureSpec.EXACTLY) {
            //Must be this size
            height = heightSize;
        } else if (heightMode == MeasureSpec.AT_MOST) {
            //Can't be bigger than...
            height = Math.min(desiredHeight, heightSize);
        } else {
            //Be whatever you want
            height = desiredHeight;
        }

        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        int width = canvas.getWidth() - getPaddingRight();
        drawChars(canvas, width);
    }

    private int drawChars(@Nullable Canvas canvas, int width) {
        int desiredHeight = getPaddingTop();

        if (mPuzzle != null) {
            Collection<Character> userChars = mPuzzle.getUserChars();
            // Compute the height that works for this width
            float offsetY = mCharH / 2;
            float offsetX = (mBoxW / 2) - (mCharW / 2);
            float x = getPaddingLeft();
            // First row
            float y = getPaddingTop() + mCharH;
            char c = 'A';
            for (int i = 0; i < 26; i++) {
                if (i % mCharsPerRow == 0) {
                    x = getPaddingLeft();
                    if (i > 0) {
                        // Another row
                        y += mCharH + offsetY;
                    }
                } else {
                    // Box width
                    x += mBoxW;
                }
                if (canvas != null) {
                    String chr = String.valueOf(c);
                    // Check if it's been mapped already
                    if (userChars.contains(c)) {
                        mTextPaint.setAlpha(96);
                    } else {
                        mTextPaint.setAlpha(255);
                    }
                    // Draw the character
                    canvas.drawText(chr, x + offsetX, y, mTextPaint);
                }
                c++;
            }
            desiredHeight = (int) (y);
        }

        desiredHeight += getPaddingBottom();
        return desiredHeight;
    }

}

File: app/src/main/java/com/pixplicity/cryptogram/views/HintView.kt
package com.pixplicity.cryptogram.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.Typeface
import android.support.v7.widget.AppCompatTextView
import android.text.TextPaint
import android.util.AttributeSet

import com.pixplicity.cryptogram.R
import com.pixplicity.cryptogram.events.PuzzleEvent
import com.pixplicity.cryptogram.models.Puzzle
import com.pixplicity.cryptogram.utils.EventProvider
import com.pixplicity.cryptogram.utils.StyleUtils
import com.squareup.otto.Subscribe


class HintView : AppCompatTextView {

    private var mPuzzle: Puzzle? = null

    private var mCharsPerRow = 1

    private var mMinBoxW: Float = 0.toFloat()
    private var mBoxW: Float = 0.toFloat()
    private var mCharH: Float = 0.toFloat()
    private var mCharW: Float = 0.toFloat()
    private var mTextPaint: TextPaint? = null


    constructor(context: Context) : super(context) {
        init(context, null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context, attrs, defStyleAttr)
    }

    private fun init(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
        val res = context.resources

        mTextPaint = TextPaint()
        mTextPaint!!.color = currentTextColor
        mTextPaint!!.isAntiAlias = true
        mTextPaint!!.typeface = Typeface.MONOSPACE

        // Compute size of each box
        mMinBoxW = StyleUtils.getSize(res, R.dimen.puzzle_box_width).toFloat()
        mTextPaint!!.textSize = StyleUtils.getSize(res, R.dimen.puzzle_hint_size).toFloat()

        // Compute size of a single char (assumes monospaced font!)
        val bounds = Rect()
        mTextPaint!!.getTextBounds("M", 0, 1, bounds)
        mCharW = bounds.width().toFloat()
        mCharH = bounds.height().toFloat()

        if (isInEditMode) {
            mPuzzle = Puzzle()
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        EventProvider.bus.register(this)
    }

    override fun onDetachedFromWindow() {
        EventProvider.bus.unregister(this)
        super.onDetachedFromWindow()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        val heightSize = MeasureSpec.getSize(heightMeasureSpec)

        val desiredWidth = suggestedMinimumWidth

        val width: Int
        //Measure Width
        if (widthMode == MeasureSpec.EXACTLY) {
            //Must be this size
            width = widthSize
        } else if (widthMode == MeasureSpec.AT_MOST) {
            //Can't be bigger than...
            width = Math.min(desiredWidth, widthSize)
        } else {
            //Be whatever you want
            width = desiredWidth
        }

        // Pack the most number of characters in the bar
        val innerBox = width - paddingLeft
        for (i in 1..25) {
            mCharsPerRow = Math.ceil((26f / i).toDouble()).toInt()
            mBoxW = (innerBox / mCharsPerRow).toFloat()
            if (mBoxW >= mMinBoxW) {
                break
            }
        }

        val desiredHeight = drawChars(null, width)

        val height: Int
        //Measure Height
        if (heightMode == MeasureSpec.EXACTLY) {
            //Must be this size
            height = heightSize
        } else if (heightMode == MeasureSpec.AT_MOST) {
            //Can't be bigger than...
            height = Math.min(desiredHeight, heightSize)
        } else {
            //Be whatever you want
            height = desiredHeight
        }

        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawChars(canvas, this.width - paddingRight)
    }

    private fun drawChars(canvas: Canvas?, width: Int): Int {
        var desiredHeight = paddingTop

        if (mPuzzle != null) {
            // Compute the height that works for this width
            val offsetY = mCharH / 2
            val offsetX = mBoxW / 2 - mCharW / 2
            var x = paddingLeft.toFloat()
            // First row
            var y = paddingTop + mCharH
            var c = 'A'
            for (i in 0..25) {
                if (i % mCharsPerRow == 0) {
                    x = paddingLeft.toFloat()
                    if (i > 0) {
                        // Another row
                        y += mCharH + offsetY
                    }
                } else {
                    // Box width
                    x += mBoxW
                }
                if (canvas != null) {
                    val chr = c.toString()
                    // Check if it's been mapped already
                    if (mPuzzle!!.isUserCharInput(c)) {
                        mTextPaint!!.alpha = 96
                    } else {
                        mTextPaint!!.alpha = 255
                    }
                    // Draw the character
                    canvas.drawText(chr, x + offsetX, y, mTextPaint!!)
                }
                c++
            }
            desiredHeight = y.toInt()
        }

        desiredHeight += paddingBottom
        return desiredHeight
    }

    @Subscribe
    fun onPuzzleProgress(event: PuzzleEvent.PuzzleProgressEvent) {
        mPuzzle = event.puzzle
        requestLayout()
    }

}
